Enh workflow (#84)
ReproNim/containers: README.md

Singularity image files have `.sing` extension.  Since we are providing
a custom filename to store the file at, we cannot guess the format of
the container (e.g., either it is
[.sif](https://www.sylabs.io/2018/03/sif-containing-your-containers/)),
so we just use uniform `.sing` extension.

# A typical YODA workflow

Lets summarize YODA principles as a possible workflow:

- create a new dataset which would contain results and everything needed
    to obtain them
- install/add subdatasets(code, other datasets, containers)
- perform the analysis using **only** materials available within the reach of this dataset.

Let's assume that our goal is to do Quality Control of an MRI dataset

> (which is available as DataLad dataset ds000003. We will create a new

Reviewer: ```suggestion
(which is available as DataLad dataset ds000003). We will create a new
```

> dataset with the output of the QC results(as analyzed by mriqc

Reviewer: ```suggestion
dataset with the output of the QC results (as analyzed by mriqc
```

BIDS-App). mriqc is provided by the ReproNim/containers dataset of
containers. Below, we execute a simple analysis workflow which 
adheres to YODA principles and we **end up with a dataset that contains

> all componentes necessary a history of how it was achieved.**

Reviewer: is spellchecker enabled?
```suggestion
all components necessary a history of how it was achieved.**
```

Reviewer: I have now added codespell workflow to this repo.  Please rebase your PR on top of master and repush -- let's see if green ;-)

This would help to guarantee reproducibility in the future because all the
materials would be *reachable* within that dataset.

## Runnable script

> For advanced users who are comfortable with datalad, the following

Reviewer: if used as a project, then
```suggestion
For advanced users who are comfortable with DataLad, the following
```
if as a command line tool
```suggestion
For advanced users who are comfortable with `datalad`, the following
```
choose the one you mean ;)

script may give you everything you need.

<details>

> <summary>The version of the script with more explanations</summary>

Author: ```suggestion
<summary>The version of the script with all commands explained</summary>
```

        --output . \
        '{inputs}' '{outputs}' participant group -w workdir
)

```
</details>

## Walkthrough

For users who are new to these components, we will walk through how
these components are used together in a typical YODA workflow.
the steps

```bash
mkdir ~/my-experiments
cd ~/my-experiments
datalad create -d ds000003-qc -c text2git
cd ds000003-qc
```

> Datalad has created a new directory for our results, `ds000003-qc`.

Reviewer: ```suggestion
DataLad has created a new directory for our results, `ds000003-qc`.
```

According to YODA principles, this dataset should also contain our input
data, code, and anything else we need to run the analysis.

Install the input dataset:

```bash
datalad install -d . -s https://github.com/ReproNim/ds000003-demo sourcedata
```

Next we install the `ReproNim/containers` collection.

```bash
datalad install -d . ///repronim/containers
```

Now let's take a look at what we have.

```bash
/ds000003-qc # The root dataset contains everything
 |--/sourcedata # we call it source, but it is actually ds000003-demo
 |--/containers # repronim/containers, this is where our non-custom code lives
```

`freeze_versions` is an optional step that will record and "freeze" the
version of the container used. Even if the `///containers` dataset is
upgraded with a newer version of our container, we are "pinned" to the

> container we explicitly determined. Note: To upgrade the container,
> rerun `freeze_versions` script with the new version.

Reviewer: ```suggestion
container we explicitly determined. Note: To switch version of the container
(e.g., to upgrade to a new one), rerun `freeze_versions` script with the version
specified.
```

Note: `freeze_versions` adds a change to the `///repronim/containers`
dataset.  Therefore, it is important to ensure that your clone of

> `///repronim/containers` is publicly available. Reproducers will use

Reviewer: I think it would be valuable to add:

```suggestion
`///repronim/containers` is publicly available and specify that URL in the `.gitmodules` instead
of the original one. Reproducers will use
```

Author: nice catch

this clone and inherit your changes to `.datalad/config`, including the
frozen versions.

```bash
datalad run -m "Downgrade/Freeze mriqc container version" \
    containers/scripts/freeze_versions bids-mriqc=0.16.0
```

When we run the bids-mriqc container, it will need a working directory

> for intermediate files. These are not helpfull to commit, so we will

Author: ```suggestion
for intermediate files. These are not helpful to commit, so we will
```
codespell is very helpfull

> tell git (and datalad) to ignore the whole directory.

Reviewer: ```suggestion
tell `git` (and `datalad`) to ignore the whole directory.
```
for being "self-consistent" ;)

        --output . \
        '{inputs}' '{outputs}' participant group -w workdir
)

```
</details>

## Walkthrough

For users who are new to these components, we will walk through how

> these components are used together in a typical Yoda workflow.

Reviewer: for consistency etc

```suggestion
these components are used together in a typical YODA workflow.
```

the steps

```bash
mkdir ~/my-experiments
cd ~/my-experiments
datalad create -d ds000003-qc -c text2git
cd ds000003-qc
```

Datalad has created a new directory for our results, `ds000003-qc`.
According to YODA principles, this dataset should also contain our input
data, code, and anything else we need to run the analysis.

Install the input dataset:

```bash
datalad install -d . -s https://github.com/ReproNim/ds000003-demo sourcedata
```

> Now we install the `ReproNim/containers` collection.

Author: ```suggestion
Next we install the `ReproNim/containers` collection.
```

```bash
datalad install -d . ///repronim/containers
```

> Now let's take a look at what we have.

Reviewer: Try to adjust avoiding starting multiple sequentially appearing sentences with `Now`.

Although I don't mind such detailed walk through -- we need to add a comment somehow/somewhere that in case of change of that monolythic script (which is "smoke tested" by CI) - we would need to adjust these steps too.

Author: Perhaps a better place for the tutorial would be on readthedocs.

IIRC, the tutorial could render code directly from the script, so we remove duplication.

Author: "Now" fixed tho

```bash
/ds000003-qc # The root dataset contains everything
    /sourcedata # we call it source, but it is actually 000003-demo

>     /repronim/containers # This is where our non-custom code lives

Reviewer: I don't think this path is correct... check

Author: Yep, this was wrong nice catch

```

Now we can "freeze" a container or containers that we plan to use so

> anyone (including you) will know exactly which version was used. This is

Reviewer: ```suggestion
anyone (including you) will know exactly which version was used and if you upgrade
the containers dataset to new version with newer versions of containers - you would need
to concisely choose to either upgrade this one or not while resolving conflicts. This is
```

technically optional, but very helpful. (All that changes is a

> subproject commit hash.)

Reviewer: I have remembered one "cons" for this IMHO worth noting. Please incorporate something like

```suggestion
subproject commit hash.)  A disadvantage though is that if you were to make
others to reproduce the results, you would need to share your clone of `///repronim/containers`
to share that updated state with frozen version in `.datalad/config`.
```

Author: Please re-read this section, I rewrote it with your ideas included.

```bash
datalad run -m "Downgrade/Freeze mriqc container version" \
    containers/scripts/freeze_versions bids-mriqc=0.16.0
```

When we run the bids-mriqc container, it will need a working directory
for intermediate files. These are not helpfull to commit, so we will
tell git (and datalad) to ignore the whole directory.

```bash
echo "workdir/" > .gitignore && datalad save -m "Ignore workdir" .gitignore
```

Now we use `datalad containers-run` to perform the analysis.

**_NOTE:_** mriqc is memory intensive, so we are restricting to a single

> participant. OOM error will cause a NodeExecutionError, and the cause

Reviewer: ```suggestion
participant. OOM (Out Of Memory) situation will cause a `NodeExecutionError`, and the cause
```
to de-crypt a bit

may not be immediately obvious.

```bash
datalad containers-run \
        --container-name containers/bids-mriqc \
        --input sourcedata \
        --output . \
        '{inputs}' '{outputs}' participant --participant-label sub-02 -w workdir
```

If everything worked as expected, we will now see our new analysis, and
a commit message of how it was obtained!

> and now you have a dataset which has a git record on how these data

Reviewer: join into above or just make it capitalized

Author: joined in and rewrote.
